Warn if PropType function is called manually

Aweary opened the pull request:
Resolves https://github.com/facebook/react/issues/7131

@gaearon so I added a new file `ReactPropTypesSecret` that is used internally to verify that a propType function is not being called manually. For the tests I basically reused what you had in #6401 as far as structure goes. Everything should be passing.

I wasn't sure if we wanted to dedupe this errors. If so I'm thinking we'd have to track it by `componentName` or something. Let me know what you think.

gaearon commented on src/isomorphic/classic/types/ReactPropTypes.js:
```diff
@@ -110,10 +111,21 @@ function createChainableTypeChecker(validate) {
     propFullName = propFullName || propName;
+    if (!__DEV__) {
+      if (secret !== ReactPropTypesSecret) {
+        console.error(
```
Let’s also check `if (typeof console !== 'undefined')`

gaearon commented on src/isomorphic/classic/types/ReactPropTypesSecret.js:
```diff
@@ -0,0 +1,17 @@
+
+'use strict';
+
+const ReactPropTypesSecret = '__REACT_PROP_TYPES_SECRET__';
```
Let’s randomize it a little bit?

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js:
```diff
@@ -0,0 +1,286 @@
+  describe('Primitive Types', function() {
+    it('should warn if called manually in production', function() {
+      spyOn(console, 'error');
+      typeCheckPass(PropTypes.array, /please/);
```
This looks misleading to me. Why are all these checks supposed to pass? They should fail, just like they fail today—the only difference is we should warn when doing a check.

Aweary replied:
Maybe I should rename `typeCheckPass`. It doesn't actually test if the validation passes or fails since that's tested elsewhere. I used valid and invalid values to be sure that the error is logged regardless of whether the validation passes or not.

Aweary replied:
Maybe something like `productionWarningCheck` instead

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js:
```diff
@@ -0,0 +1,286 @@
+    );
+    const expectedCount = __DEV__ ? 0 : 1;
+    expect(console.error.calls.count()).toBe(expectedCount);
+    if (!__DEV__) {
```
Doesn’t this test run only in prod mode? Why do we check `__DEV__`?

Aweary replied:
There's a test towards the bottom that tests that these warnings don't appear in development mode.

gaearon replied:
I think it is unnecessary. Our tests are set up in a way that any unexpected warning would fail a test. So if it showed up anywhere, we’d notice thanks to existing DEV mode tests. Let’s make these tests production-only. This simplifies the logic too.

Aweary replied:
Cool that works for me 👍

gaearon commented on src/isomorphic/classic/types/ReactPropTypes.js:
```diff
@@ -110,10 +111,21 @@ function createChainableTypeChecker(validate) {
     propFullName = propFullName || propName;
+    if (!__DEV__) {
+      if (secret !== ReactPropTypesSecret && typeof console !== 'undefined') {
+        console.error(
```
Let’s keep a cache to avoid spamming the console. I think logging this warning just once should be enough.

Aweary replied:
Just once per what, the type checker or the prop/component?

gaearon replied:
I would just log it once at all. Logging can get expensive in production (depends on the browser).

Aweary replied:
So should I use a generic message that doesn't mention the component name/prop then? I included that because I thought it would be useful, but if we're only warning once globally then it won't really matter

gaearon replied:
I think mentioning the component name and prop might still be useful—you want to have at least a vague idea of where it’s happening.

Aweary replied:
My concern was that if we only warned once and they are calling PropType functions in multiple places they are only being warned about one of those instances. That might be annoying, especially in production, since you'd think you fixed it (since you only got one warning) and then you'd get the warning again for the next instance.

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js:
```diff
@@ -0,0 +1,286 @@
+      'testComponent',
+      'prop'
+    );
+    const expectedCount = __DEV__ ? 0 : 1;
```
Let’s make an `if` branch and do different assertions in `if` and `else`. This would read simpler IMO.

Aweary replied:
Well if we're not doing any `__DEV__` tests then it would just be a constant `0` right?

gaearon commented on src/isomorphic/classic/types/ReactPropTypesSecret.js:
```diff
@@ -0,0 +1,17 @@
+
+'use strict';
+
+const ReactPropTypesSecret = '__REACT_PROP_TYPES_SECRET__' + Math.random();
```
Nit: `.toString()`.

Aweary replied:
👍  will do. Just curious, why? It'll be coerced to a string either way AFAIK.

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js:
```diff
@@ -0,0 +1,253 @@
+    global.process = oldProcess;
+  });
+
+  function productionWarningCheck(declaration, value) {
```
Let’s call it `expectWarningInProduction`.

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js:
```diff
@@ -0,0 +1,253 @@
+    });
+  });
+
+  it('should not warn when doing type checks internally', function() {
```
Maybe `should not warn for type checks performed by React`

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js:
```diff
@@ -0,0 +1,253 @@
+      <Component foo="foo" bar={42} baz={{ qux: [1, 2, 3] }}/>
+    );
+    instance = ReactTestUtils.renderIntoDocument(instance);
+    expect(console.error.calls.count()).toBe(0);
```
Let’s also check we don’t emit the unnecessary warning even if the check fails?

gaearon commented:
>My concern was that if we only warned once and they are calling PropType functions in multiple places they are only being warned about one of those instances. That might be annoying, especially in production, since you'd think you fixed it (since you only got one warning) and then you'd get the warning again for the next instance.

Now that I think of it, why not warn in development too? It would be much easier to catch that issue even if the warning is only printed once.

Sorry about the churn 😄 

So we could do the following:

* Warn once on any manual call in development
* Don’t do anything in production

Do you think this makes more sense?
We’d still change *production* behavior eventually, but at least people would be aware that calling them manually is not OK during development.

Aweary commented:
> Now that I think of it, why not warn in development instead? It would be much easier to catch that issue even if the warning is only printed once.

I agree, production-only errors might be frustrating since some users might not catch those until it's deployed to a staging environment (or even to actual production). 

> Sorry about the churn 😄

No worries at all! I'd rather get it done right than get it done quick 😄 

> So we could do the following:
* Warn once on any manual call in development
* Don’t do anything in production

That seems sensible and consistent with other warnings/errors in React. Since we'd be doing these warnings in development, I think it would make the most sense to warn once for every instance where they're calling PropTypes function manually. Since dev warnings aren't really expensive, I feel like that would be the most useful. 

All we'd need to do is create a cache in the `createChainableTypeChecker` closure that stores the component and prop name (like `{ component { prop } }`) and check against that.

Does that sound good @gaearon?

gaearon commented:
>I think it would make the most sense to warn once for every instance where they're calling PropTypes function manually.

I’m hesitant because I can’t imagine somebody calling them as one-offs. If somebody actually calls those functions manually, I imagine they might be a part of something like a form validator, in which case you’d see a warning on every keystroke.

Do you see deduping based on `componentName:propName` a good compromise?

Aweary commented:
> Do you see deduping based on componentName:propName a good compromise?

That's actually what I was trying to recommend. Checking if a warning has been logged for  `${component}:${prop}` so that it's not logged multiple times if props are checked again, but it still logs once for every instance where they are manually calling the PropType function. 

By "instance" I meant every place where they're manually calling a PropType function, not every time the function is invoked. Sorry, my terminology was probably a little confusing.

gaearon commented:
Well, technically, you can’t really know *where* it’s being called from. You only know what arguments are being passed. But seems like we agree on the proposal.

Aweary commented:
@gaearon alright, so I updated it so that it warns in dev instead of production. In dev it creates a cache in `createChainableTypeChecker` so each cache is unique to each PropType function and if the internal secret isn't provided it creates a cache key and checks the cache before warning.

I also had to update a few other files that called PropType functions manually so that all the tests would pass.

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypesProduction-test.js:
```diff
@@ -0,0 +1,262 @@
+'use strict';
+
+
+describe('ReactPropTypesProduction', function() {
```
Looks like we can delete this file now and instead add a simple test against unguarded calll to the DEV file?

Aweary replied:
Yeah definitely!

gaearon commented on src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js:
```diff
@@ -40,7 +43,9 @@ function typeCheckPass(declaration, value) {
-    ReactPropTypeLocations.prop
+    ReactPropTypeLocations.prop,
+    null,
+    ReactPropTypesSecret
```
I’m a bit concerned about this. We’re not actually testing that *React* is passing the secret when validating. So if React doesn’t do it in some case, we’ll miss this.

Can we change these `typeCheckFail` and `typeCheckPass` to be more like integration tests? i.e. they would create a class and then make sure that calling `createElement` produces (or doesn’t produce) the relevant warning. Or am I missing something?

Aweary replied:
> I’m a bit concerned about this. We’re not actually testing that React is passing the secret when validating.

There is a test in `ReactPropTypesProduction-test.js` that makes sure the warnings aren't being called when prop types are validated internally.

> So if React doesn’t do it in some case.

React should do it all cases, I updated all the files that called a PropType function internally, and I would say it's reasonable that any future changes that involve calling one of these PropType functions should also be sure to pass in `ReactPropTypeSecretValue`. 

> we’ll miss this.

It's hard to miss, the test suite blows up if a warning is logged when it shouldn't be.

> Can we change these typeCheckFail and typeCheckPass to be more like integration tests? i.e. they would create a class and then make sure that calling createElement produces (or doesn’t produce) the relevant warning. Or am I missing something?

I can definitely do that. As is I don't think it makes a difference, but it would ensure that future changes to how PropTypes are checked don't require this test to be updated as well.

The only issue I see is that if we do it this way we can't get the actual error instance returned from `declaration` so it would have to rely on spying on `console.error` instead.

gaearon replied:
>It's hard to miss, the test suite blows up if a warning is logged when it shouldn't be.

Yea, you’re right, I just tried and existing `ReactPropTypes-test`s catch that.

Aweary replied:
I learned that the hard way once everything started failing after moving this from production to dev 😄

gaearon commented:
OK, this looks good, let’s just remove `Production` tests now since they’re no longer relevant.
Thanks!

Aweary commented:
@gaearon I removed `ReactPropTypesProduction-tests.js` and moved those tests to the corresponding blocks in `ReactPropTypes-tests.js`, all tests passing!

gaearon commented on src/isomorphic/classic/types/ReactPropTypes.js:
```diff
@@ -104,16 +106,39 @@ function is(x, y) {
+            false,
+            'You are manually calling a React.PropTypes validation ' +
+            'function for the prop %s on %s. This is deprecated ' +
+            'and will not work in the next major version.',
```
Let’s change this to be

```js
'You are manually calling a React.PropTypes validation ' +
'function for the `%s` prop on `%s`. This is deprecated ' +
'and will not work in the next major version.',
```

gaearon replied:
Updated the above message: please include the link. (It doesn’t actually work yet but I’ll fix it tomorrow). We need to explain that [some third-party proptypes](https://github.com/react-bootstrap/react-prop-types/blob/master/src/all.js) might need to be updated to pass `...rest` arguments down now, or they would trigger a false positive.

gaearon commented:
I ran this on [React Bootstrap docs page](http://react-bootstrap.github.io/components.html) and I see a lot of warnings related to React Router:

<img width="1165" alt="screen shot 2016-06-30 at 02 08 08" src="https://cloud.githubusercontent.com/assets/810438/16473695/b1a37b9c-3e67-11e6-84b7-7da4fb94c21f.png">

Apparently it is [checking prop types for you](https://github.com/reactjs/react-router/blob/13f327be99baa504e3e10343aa69b256af59da65/modules/RouteUtils.js#L12-L24). I’m not sure I fully understand the intention there—what would this catch that creating an element wouldn’t?

gaearon commented:
To clarify my above comment, amending the warning message is the only thing holding it back.
(If you’d also like to write the warning gist, feel free to!)

Aweary commented:
@gaearon I updated the message to include the link! 

>(If you’d also like to write the warning gist, feel free to!) 
I gave it a shot https://gist.github.com/Aweary/d8bdcaaa410ea1daa4b541c527c0131a. I'm not sure if it's what you guys are looking for specifically, but if so great! If not, I'm happy to make any changes if you'd like to use the bulk of it.

Thanks for all the feedback on this, it's been fun!

gaearon commented:
Thanks, looks good to me. I’ll merge after putting up the gist (yours is a good start!)

mjackson commented:
We used to run `propTypes` checks because I didn't understand how React did it for us inside `createElement`. For some reason I thought that if we didn't actually use the element (`<Route>` "elements" have always been non-renderable, a frequent point of contention and something we plan on fixing shortly) that those checks wouldn't run. I was probably wrong.

We don't actually do this in current versions of the router, so this shouldn't be a problem.

gaearon commented:
React wouldn't create it at route definition time (unless you call something like `createElement(route.component, route)`) but presumably the element would get created when you want to render something. So I think the only downside is that props in route config would not get validated until that particular route gets rendered. Which doesn't seem like a big deal anyway.

mjackson commented:
I'm talking about the creation of the actual `Route` component, not its `<Route component>`. We were running checks to make sure the props you gave your `<Route>` were valid. At that point, you've already done a `createElement(Route, ...)`.

gaearon commented:
Hmm. This should be covered by whenever you `createElement` to actually render it.

gaearon commented:
Do you mind rebasing and fixing the tests? Seems like the new oneOfType/oneOf warnings cause tests to fail.

Aweary commented:
@gaearon so looking at the new union type warning check, it looks like it returns early so this warning would never be logged 

```js
  if (!Array.isArray(arrayOfTypeCheckers)) {
    warning(false, 'Invalid argument supplied to oneOfType, expected an instance of array.');
    return emptyFunction.thatReturnsNull;
  }
```

Do you want both warnings to be logged if they are both applicable? Or should I just adjust the tests to not expect this warning if they're using invalid arguments?

gaearon commented:
>Or should I just adjust the tests to not expect this warning if they're using invalid arguments?

Yea, this would be fine. (It’s quite a corner case.)

gaearon commented:
Ping 😉 
Better to get this in sooner rather than later so it doesn’t get more conflicts.

Aweary commented:
Sorry, holiday weekend (🇺🇸🎉), I'll try and find a moment this afternoon!

gaearon commented:
Oh man, I totally forgot about this being a holiday in US, sorry!
No rush at all.

Aweary commented:
@gaearon I removed the tests that expected the warning for those union types with invalid args 👍

gaearon commented:
Thank you!

Aweary commented:
🎉  thanks for all the feedback/help!

spicyj commented:
I'm a little concerned that this will now warn when defining components with a different copy of React than the renderer comes from. Instead of randomizing the secret can we just hardcode something?

gaearon commented:
Ah, good catch.
Let’s change to `SECRET_DO_NOT_PASS_THIS_EXPLICITLY_OR_YOU_WILL_BE_FIRED`.

gaearon commented:
@Aweary Would you like to make a followup PR?

spicyj commented:
Lol we don't need to fire people for everything they shouldn't do! `'react.proptypes.call.secret'` or something would be just fine too. :)

Aweary commented:
@gaearon I'd be happy to! I'll get a PR going right now

jedwards1211 commented:
@spicyj but there were plenty of other great possibilities! like `SECRET_DO_NOT_PASS_THIS_EXPLICITLY_OR_YOU_WILL_BE_DEFENESTRATED`.  I always wish there were more opportunities for jokes embedded in code.
